### PR TITLE
Lab07 - Errors

### LINKED ISSUE #530: Lab07 - Errors
**Create errors for:**
- Value < 0
- ID not found in Read, Update and Delete
- Add locking for proper use of sync.Map

### PR DESCRIPTION
**Pull request by willshen8**
Fixes #530

Review of colleague's PR #533

## Creating error conditions for the following two conditions
- Value < 0
- ID not found in Read, Update and Delete
Also adding locking for cater for race conditions

### CODE REVIEW

File: 07_errors/willshen8/syncStore.go
Comment (pfannerj) [inline 07_errors/willshen8/syncStore.go]: The example provided by Dan for lab6 uses `s.mu.Lock()` followed directly by `defer s.mu.Unlock()` at the beginning of the each applicable function.  I believe using defer is good practice because then it doesn't matter where the function exits, the defer means it will always unlock before the function exits.

File: 07_errors/willshen8/syncStore.go
Comment (pfannerj) [inline 07_errors/willshen8/syncStore.go]: My question here is whether or not the Lock should be done before the Load... in the example that @anzboi  provided with lab6, it does the Lock as the first statement and then defer Unlock (as mentioned in my comment above)... I see ur logic in that u are only locking if u need to update but I think using the Lock/defer Unlock pattern is preferable.

File: 07_errors/willshen8/error.go
Comment (pfannerj) [inline 07_errors/willshen8/error.go:13]: I like the idea in the example where the error codes are defined as an enum within error.go and then the more meaningful name (eg. ErrNotFound) is passed into the error function rather than just passing a 1 or 2.

**Comment by pfannerj** [comment]
Everything LGTM but I'll hold off labelling with "Approved by colleague" until I confirm that it's ok to Lock after the Load... or maybe you wanna just make that change and use defer Unlock?? 
Also noticed that your not looking for the bonus points for the build the new leveldb storer...

**Comment by pfannerj** [comment]
I just had a chat with Dan and he says the update should definitely do the Lock and defer Unlock up front at the start of the function.. so I'll mark with 'Changes Requested' and just let me know when ur ready for me to have another look...

File: 07_errors/willshen8/syncStore.go
Comment (willshen8) [reply 07_errors/willshen8/syncStore.go]: Thanks, changed added.

File: 07_errors/willshen8/syncStore.go
Comment (willshen8) [reply 07_errors/willshen8/syncStore.go]: II have tried locking it before the  read operation and it goes to an infinite loop.

File: 07_errors/willshen8/error.go
Comment (willshen8) [reply 07_errors/willshen8/error.go:13]: I was thinking about this one and thought what they asking is an error code similar to network response code like 404, 200, etc. 

The one in the lecture slides `var ErrPermission = errors.New("permission denied")` actually returns a string. And my idea of code is a number, but I could be wrong.

So my change is to add the equivalent of enum:
```
const (
	NEGATIVEVALUE = 1001
	NONEXISTENTPUPPY = 1002
)
```

**Comment by willshen8** [comment]
@pfannerj Thanks judy, changes added after your PR review. Please approve if you are happy with it,

File: 07_errors/willshen8/error.go
Comment (pfannerj) [reply 07_errors/willshen8/error.go:13]: yeah, that's good, numeric code is in the example too.

File: 07_errors/willshen8/syncStore.go
Comment (pfannerj) [reply 07_errors/willshen8/syncStore.go]: hmmm, works fine for me... 
```
// UpdatePuppy modifies puppy data in the sync store, either creating a new one or overwriting an old one.
func (s *SyncStore) UpdatePuppy(puppyID uint32, p *Puppy) (uint32, error) {
	s.Lock()
	defer s.Unlock()
	if _, ok := s.Load(puppyID); !ok {
		s.currID++
		puppyID = s.currID
	}
	p.ID = puppyID //ensure the ID within p always matches the sync store key (puppyID)
	s.Store(puppyID, *p)
	return puppyID, nil
}
```
Do u want me to come and visit u?

Comment (pfannerj) [reply 07_errors/willshen8/syncStore.go]: Sorry... I don't know how to insert code so that it displays properly

Comment (camscale) [reply 07_errors/willshen8/syncStore.go]: @pfannerj Use triple backticks around the block of code instead of single backticks. I've edited your comment, so take a look

Comment (willshen8) [reply 07_errors/willshen8/syncStore.go]: @camh-anz  Thanks Cam, I told her the answer via slack yesterday.

File: 07_errors/willshen8/error.go
Comment (anzboi) [inline 07_errors/willshen8/error.go]: Go convention is to name these the same as any other variable
`NegativeValue`
`NonExistentPuppy`

File: 07_errors/willshen8/storer_test.go
Comment (anzboi) [inline 07_errors/willshen8/storer_test.go]: You can use `assert.EqualValues(t, expected, actual)` to check the errors directly, rather than printing, capturing output and checking the error strings

Note that your interfaces return a generic `error` rather than your `*Error`. I actually prefer this way of doing it, but you may need to type cast the returned error first (though I think `EqualValues` handles this via reflection anyway). If you do need to, here's the code

```go
createID, createError := s.storer.CreatePuppy(&s.InvalidPuppy)

// You may not need these lines if EqualValues works on its own
// require.IsType(t, *Error{}, createError) // Check the type is correct, fail and stop test if not
// err := createError.(*Error)

assert.EqualValues(t, expected, createError)
```

File: 07_errors/willshen8/storer_test.go
Comment (anzboi) [inline 07_errors/willshen8/storer_test.go]: Store these as unexported variables
```
var (
  puppy1 = Puppy{1, "Jack Russell Terrier", "White and Brown", "1500"}
  ...

  createError = Error{Code: NEGATIVEVALUE, Message: "Puppy value can't be less than 0"}
  ...
)
```

This also avoids recreating them for every test (just don't modify them in any tests, copy first if you need to modify)

File: 07_errors/willshen8/storer_test.go
Comment (anzboi) [inline 07_errors/willshen8/storer_test.go]: Move these to the variables at the top of the file.

File: 07_errors/willshen8/puppy.go
Comment (anzboi) [inline 07_errors/willshen8/puppy.go:15]: Not sure if value should be numeric or a string. Most days I would argue numeric, but there are non-trivial challenges storing monetary data, and for the json api you will need to do the integer conversion anyway so I will let this slide.

On a side note, A couple of people at anz have been working on an IEEE decimal implementation in go (which solves the problems with storing dollar values). Once that is in, there would be no reason to keep using strings for dollar values.

**Review by anzboi — CHANGES_REQUESTED** [review]
Looking good, love the test coverage.
I would like to rework the suite tests though, move the test data out of the suite to unexported variables, and rework how you check your errors (check them directly rather than through output of `Error()`)

File: 07_errors/willshen8/error.go
Comment (willshen8) [reply 07_errors/willshen8/error.go]: Thanks agreed, was thinking Java when I wrote these constants.

File: 07_errors/willshen8/storer_test.go
Comment (willshen8) [reply 07_errors/willshen8/storer_test.go]: Thanks, this is addressed. I simply used the testingSuite `Equal` function.

File: 07_errors/willshen8/storer_test.go
Comment (willshen8) [reply 07_errors/willshen8/storer_test.go]: Addressed also.

File: 07_errors/willshen8/storer_test.go
Comment (willshen8) [reply 07_errors/willshen8/storer_test.go]: Done!

File: 07_errors/willshen8/puppy.go
Comment (willshen8) [reply 07_errors/willshen8/puppy.go:15]: I think the original intention was that we could have currency signs like $.

File: 07_errors/willshen8/storer_test.go
Comment (willshen8) [reply 07_errors/willshen8/storer_test.go]: Hi Dan, please note that I still used this old-style for one of the test cases for 100% test coverage purpose.

**Review by anzboi — APPROVED** [review]
LGTM, great work

**Pull request merged**
